### ISSUE TITLE
The performance stress test results of EAGLE-3 are not good

### ISSUE THREAD

**Issue by jason-i-vv** [sgl-project/sglang#5274]
I used EvalScope to stress test the inference performance of SGLang. A very positive aspect is that SGLang outperforms vLLM. However, when I tried to enable the EAGLE-3 optimization, the throughput dropped sharply.

My environment is an NVIDIA A6000, and the stress test command is as follows:
```
docker run --rm --net=host \
--mount type=bind,source=/data,target=/data \
registry.bingosoft.net/bingomatrix/modelscope:ubuntu22.04-cuda12.1.0-py310-torch2.4.0-tf2.16.1-1.18.1-vllm-evalscope \
evalscope perf \
  --parallel 20 \
  --model /data/models/Meta-Llama-3.1-8B-Instruct \
  --url http://127.0.0.1:30000/v1/chat/completions \
  --api openai \
  --dataset random \
  --min-tokens 128 \
  --max-tokens 128 \
  --prefix-length 64 \
  --min-prompt-length 1024 \
  --max-prompt-length 2048 \
  --number 100 \
  --tokenizer-path Qwen/Qwen2.5-0.5B-Instruct \
  --debug
```

Start SGLang without EAGLE-3:
```
docker run --gpus all \
    --shm-size 32g \
    -p 30000:30000 \
    -v /data/:/data \
    --ipc=host \
    registry.bingosoft.net/bingomatrix/sglang:v0.4.5-cu125 \
    python3 -m sglang.launch_server --model-path /data/models/Meta-Llama-3.1-8B-Instruct --host 0.0.0.0 --port 30000 \
    --mem-fraction 0.6 \
    --cuda-graph-max-bs 2 --dtype float16
```
And Test Results is: 
```
Benchmarking summary:
+-----------------------------------+----------------------------------------------------------------------+
| Key                               | Value                                                                |
+===================================+======================================================================+
| Time taken for tests (s)          | 58.761                                                               |
+-----------------------------------+----------------------------------------------------------------------+
| Number of concurrency             | 20                                                                   |
+-----------------------------------+----------------------------------------------------------------------+
| Total requests                    | 100                                                                  |
+-----------------------------------+----------------------------------------------------------------------+
| Succeed requests                  | 100                                                                  |
+-----------------------------------+----------------------------------------------------------------------+
| Failed requests                   | 0                                                                    |
+-----------------------------------+----------------------------------------------------------------------+
| Throughput(average tokens/s)      | 217.833                                                              |
+-----------------------------------+----------------------------------------------------------------------+
| Average QPS                       | 1.702                                                                |
+-----------------------------------+----------------------------------------------------------------------+
| Average latency (s)               | 11.74                                                                |
+-----------------------------------+----------------------------------------------------------------------+
| Average time to first token (s)   | 11.74                                                                |
+-----------------------------------+----------------------------------------------------------------------+
| Average time per output token (s) | 0.092                                                                |
+-----------------------------------+----------------------------------------------------------------------+
| Average input tokens per request  | 2489.74                                                              |
+-----------------------------------+----------------------------------------------------------------------+
| Average output tokens per request | 128.0                                                                |
+-----------------------------------+----------------------------------------------------------------------+
| Average package latency (s)       | 11.74                                                                |
+-----------------------------------+----------------------------------------------------------------------+
| Average package per request       | 1.0                                                                  |
+-----------------------------------+----------------------------------------------------------------------+
| Expected number of requests       | 100                                                                  |
+-----------------------------------+----------------------------------------------------------------------+
| Result DB path                    | outputs/20250411_111326/Meta-Llama-3.1-8B-Instruct/benchmark_data.db
```
Start SGLang with EAGLE-3:
```
docker run --gpus all \
    --shm-size 32g \
    -p 30000:30000 \
    -v /data/:/data \
    --ipc=host \
    registry.bingosoft.net/bingomatrix/sglang:v0.4.5-cu125 \
    python3 -m sglang.launch_server --model-path /data/models/Meta-Llama-3.1-8B-Instruct --host 0.0.0.0 --port 30000 \
    --speculative-algorithm EAGLE3 \
    --speculative-draft-model-path /data/models/sglang-EAGLE3-Llama-3.1-Instruct-8B \
    --speculative-num-steps 5 \
    --speculative-eagle-topk 8 --speculative-num-draft-tokens 32 --mem-fraction 0.6 \
    --cuda-graph-max-bs 16 --dtype float16
```

Test results is:
```
Benchmarking summary:
+-----------------------------------+----------------------------------------------------------------------+
| Key                               | Value                                                                |
+===================================+======================================================================+
| Time taken for tests (s)          | 77.861                                                               |
+-----------------------------------+----------------------------------------------------------------------+
| Number of concurrency             | 20                                                                   |
+-----------------------------------+----------------------------------------------------------------------+
| Total requests                    | 100                                                                  |
+-----------------------------------+----------------------------------------------------------------------+
| Succeed requests                  | 100                                                                  |
+-----------------------------------+----------------------------------------------------------------------+
| Failed requests                   | 0                                                                    |
+-----------------------------------+----------------------------------------------------------------------+
| Throughput(average tokens/s)      | 164.395                                                              |
+-----------------------------------+----------------------------------------------------------------------+
| Average QPS                       | 1.284                                                                |
+-----------------------------------+----------------------------------------------------------------------+
| Average latency (s)               | 14.604                                                               |
+-----------------------------------+----------------------------------------------------------------------+
| Average time to first token (s)   | 14.604                                                               |
+-----------------------------------+----------------------------------------------------------------------+
| Average time per output token (s) | 0.114                                                                |
+-----------------------------------+----------------------------------------------------------------------+
| Average input tokens per request  | 2489.74                                                              |
+-----------------------------------+----------------------------------------------------------------------+
| Average output tokens per request | 128.0                                                                |
+-----------------------------------+----------------------------------------------------------------------+
| Average package latency (s)       | 14.604                                                               |
+-----------------------------------+----------------------------------------------------------------------+
| Average package per request       | 1.0                                                                  |
+-----------------------------------+----------------------------------------------------------------------+
| Expected number of requests       | 100                                                                  |
+-----------------------------------+----------------------------------------------------------------------+
```

So, When using eagle-3, the performance becomes worse, what is the reason?

**Comment by Fridge003** [comment]
Speculative decoding is for small batch sizes rather than stress test.  When the batch size is large, the overhead and extra memory will negatively affect performance.

**Comment by jason-i-vv** [comment]
<img width="823" alt="Image" src="https://github.com/user-attachments/assets/8c88f921-9e8e-4eec-875e-1dd08637a81c" />

So,How can I get this performance test? Are there any test documents . Thanks!

**Comment by Fridge003** [comment]
You can try this file for testing speculative decoding
https://github.com/sgl-project/sglang/blob/main/python/sglang/test/send_one.py

**Comment by jason-i-vv** [comment]
got it. i will try .

**Comment by jason-i-vv** [comment]
> You can try this file for testing speculative decoding https://github.com/sgl-project/sglang/blob/main/python/sglang/test/send_one.py

In this case, speculative decoding resulted in over a 2x performance improvement.

```
# without speculative decoding
speed=42.00 token/s

# with speculative decoding
speed=92.47 token/s
```

**Comment by Tekgnosis** [comment]
I'm using SGLang on RTX 3090 24GB to run Qwen3 8b Q6_K_L and attempting to use the EAGLE3 trained for this model.  I found this bug in a search because I'm having a heck of a time getting EAGLE3 to actually improve on the baseline.  I'm NOT using any batching here, this is a single 500 token output request in each case.  My assumption here is that the EAGLE3 training process is possibly not very good yet for QWEN3.  I did have to convert the EAGLE3 model from b16 to f16 to use it or the dtype would mismatch because I'm using AWQ for the main model. 

I used the send_one.py noted above:
  Baseline:
  - acc_length = 1.00 (no speculation)
  - speed = 125.06 t/s

  EAGLE3 (optimal parameters):
  - acc_length = 3.10 ← Above the threshold noted for what should result in benefits exceeding overhead.
  - speed = 115.31 t/s (-7.8% vs baseline)

So, even though I'm getting >3 accepted tokens speculated per pass, we're still below baseline.

Issues:  overlap scheduler is disabled when EAGLE3 is active.  Batch size is limited to 2 vs 16 on my baseline.

My initial results and configuration, which might be valuable for anyone attempting to get sglang working on sm86:

  Configuration Performance:
  - Baseline (no EAGLE3): 133.8 t/s (not sure why this is more than the send_one.py but it's my daily)
  - EAGLE3 Low Overhead (steps=3, topk=2, draft=4): 105.2 t/s (-21.4%)
  - EAGLE3 Moderate (steps=4, topk=5, draft=16): 104.5 t/s (-21.9%)
  - EAGLE3 Default (steps=6, topk=10, draft=32): 112.2 t/s (-16.2%) ← BEST
  - EAGLE3 Tuned (steps=6, topk=5, draft=6): 101.9 t/s (-23.8%)
  - EAGLE3 FlashInfer (default + flashinfer backend): 109.2 t/s (-18.4%)

System Configuration

  OS: Ubuntu 24.04.2 LTS
  Kernel: 6.6.87.2-microsoft-standard-WSL2 (WSL2)
  Python: 3.12.3
  GPU: NVIDIA GeForce RTX 3090 (24GB VRAM)
  CUDA Driver: 581.42
  CUDA Toolkit: 12.9 (release 12.9, V12.9.86)

  Python Dependencies

  SGLang: 0.5.3.post1
  PyTorch: 2.8.0+cu128
  Triton: 3.4.0
  FlashInfer: 0.4.0
  XGrammar: 0.1.25
  vLLM: 0.11.0
  TorchAO: 0.9.0

  Models

  Base Model:
  Name: Qwen3-8B-AWQ
  Source: HuggingFace (AWQ quantized variant)
  Model Type: qwen3
  Vocab Size: 151,936 tokens
  Quantization: AWQ 4-bit (group_size=128, backend=autoawq, version=gemm)
  Dtype: float16
  Size on Disk: 5.7 GB
  Architecture: Qwen3ForCausalLM

  EAGLE3 Draft Model:
  Name: Qwen3-8B-EAGLE3
  Source: Tengyunw/qwen3_8b_eagle3 (HuggingFace)
  Architecture: LlamaForCausalLMEagle3
  Model Type: llama (base architecture)
  Vocab Size: 151,936 tokens (target) + 32,000 (draft_vocab_size)
  Dtype: float16 (modified from bfloat16 in config.json)
  Size on Disk: 1.5 GB
  Note: Uses LlamaForCausalLMEagle3 wrapper with Qwen3 vocabulary

Just here to add the data in case anyone else is doing something similar.
